x/delayedack/keeper/finalize.go

```diff
@@ -3,36 +3,79 @@ package keeper
+func (k Keeper) FinalizeRollappPackets(ctx sdk.Context, ibc porttypes.IBCModule, rollappID string, stateEndHeight uint64) (int, error) {
+	// Get the latest state info of the rollapp
+	latestIndex, found := k.rollappKeeper.GetLatestFinalizedStateIndex(ctx, rollappID)
+	if !found {
+		return 0, fmt.Errorf("latest finalized state index for rollapp '%s' is not found", rollappID)
+	}
+	stateInfo, found := k.rollappKeeper.GetStateInfo(ctx, rollappID, latestIndex.Index)
+	if !found {
+		return 0, fmt.Errorf("stateInfo for rollapp '%s' is not found", rollappID)
+	}
+
+	// Check the latest finalized height of the rollapp is higher than the height specified
+	if stateEndHeight > stateInfo.GetLatestHeight() {
+		return 0, fmt.Errorf("packet height '%d' is not finalized yet, latest height '%d'", stateEndHeight, stateInfo.GetLatestHeight())
+	}
+
+	// Get all pending rollapp packets until the specified height
 	rollappPendingPackets := k.ListRollappPackets(ctx, types.PendingByRollappIDByMaxHeight(rollappID, stateEndHeight))
 	if len(rollappPendingPackets) == 0 {
-		return nil
+		return 0, nil
 	}
-	logger := ctx.Logger().With("module", "DelayedAckMiddleware")
-	// Get the packets for the rollapp until height
-	logger.Debug("finalizing IBC rollapp packets",
-		"rollappID", rollappID,
-		"state end height", stateEndHeight,
-		"num packets", len(rollappPendingPackets))
-	for _, rollappPacket := range rollappPendingPackets {
-		if err := k.finalizeRollappPacket(ctx, ibc, rollappID, logger, rollappPacket); err != nil {
-			return fmt.Errorf("finalize rollapp packet: %w", err)
+
+	// Finalize the packets
+	for _, packet := range rollappPendingPackets {
+		if err := k.finalizeRollappPacket(ctx, ibc, rollappID, packet); err != nil {
+			return 0, fmt.Errorf("finalize rollapp '%s' packet: %w", rollappID, err)
 		}
 	}
+
+	return len(rollappPendingPackets), nil
+}
+
+// FinalizeRollappPacket finalizes a singe packet by its rollapp packet key.
+func (k Keeper) FinalizeRollappPacket(ctx sdk.Context, ibc porttypes.IBCModule, rollappID string, rollappPacketKey string) error {
+	// Get a rollapp packet
+	packet, err := k.GetRollappPacket(ctx, rollappPacketKey)
+	if err != nil {
+		return fmt.Errorf("get rollapp packet: %w", err)
+	}
+
+	// Get the latest state info of the rollapp
+	latestIndex, found := k.rollappKeeper.GetLatestFinalizedStateIndex(ctx, rollappID)
+	if !found {
+		return fmt.Errorf("latest finalized state index for rollapp %s is not found", rollappID)
+	}
+	stateInfo, found := k.rollappKeeper.GetStateInfo(ctx, rollappID, latestIndex.Index)
+	if !found {
+		return fmt.Errorf("stateInfo for rollapp %s is not found", rollappID)
+	}
+
+	// Check the latest finalized height of the rollapp is higher than the packet proof height
+	if packet.ProofHeight > stateInfo.GetLatestHeight() {
+		return fmt.Errorf("packet height '%d' is not finalized yet, latest height '%d'", packet.ProofHeight, stateInfo.GetLatestHeight())
+	}
```

Comment:
This seems to be more general purpose 'is height finalized' function
Can you extract it?